core: initial encoding conversion support

jgmdev opened the pull request:
New dependency introduced:
* uchardet - for automatic encoding detection

This PR is for performing automatic encoding detection and conversion on the fly from non UTF-8 to UTF-8 when opening files and from UTF-8 to non UTF-8 on file save. 

Also it properly handles byte order marks management on UTF-16LE, UTF-16BE, UTF-32LE. UTF-32BE, UTF-7, UTF-8 and GB18030 when opening and saving, it also uses the BOM to detect the file charset on file load.

Already in working shape but needs testing and maybe improvements (one of them having a uchardet.wrap for meson build system to support static building), leaving it here for those interested.

![loading-non-utf-8-encodings](https://user-images.githubusercontent.com/1702572/196560029-89d95acb-b5aa-4db3-b4f2-8bece94fa009.png)

![select-output-encoding-on-save](https://user-images.githubusercontent.com/1702572/196560037-8dab6e10-5cd3-4384-a5b4-c505f1b37879.png)

adamharrison commented:
So, just to repeat what I mentioned on discord: I'm not hugely in favour of expanding core with APIs that are probably not going to be used very much. As it stands, I would think most people work in UTF-8 these days, which is what we are using internally in the editor.

Converting between encodings is great; but I think it's probably better as a plugin. I could be wrong (maybe it's a reflection of my problem domain specifically that I work in); but almost never in my day-to-day do I swap between encodings, or interact with anything that isn't ASCII or UTF-8; and it's only getting less likely as time goes by. (I think something like <1% of websites these days are not ASCII or UTF-8).

So while I think it's good as a specialist plugin, I don't think we necessarily need to have it in core.

jgmdev commented:
> Converting between encodings is great; but I think it's probably better as a plugin. I could be wrong (maybe it's a reflection of my problem domain specifically that I work in); but almost never in my day-to-day do I swap between encodings, or interact with anything that isn't ASCII or UTF-8; and it's only getting less likely as time goes by. (I think something like <1% of websites these days are not ASCII or UTF-8).

This isn't  solely about the ability to convert to other encodings/charsets, that just happen to be a happy coincidence. It is about been able to properly read files that don't contain UTF-8/ASCII charsets. I have experienced several times not been able to open some downloaded text files or old source files I worked on the past and having to resort to other text editors for editing/conversion which makes lite-xl look like a joke or toy text editor. 

A text editor should be able to handle to the best of its capacities opening text files for editing, and I'm not the only one who have experienced the aforementioned issues. I remember some others on discord trying to open non utf-8 files on lite-xl and failing even with crashes or multiple issues due to the inconsistencies.

I don't agree that this shouldn't be part of the core and just a separate plugin, it is an essential functionality found in every major text editor. Most users expect their editor to be able to handle most text files thrown to it and this PR targets that. Depending on other text editors to handle non UTF-8 isn't great, if this wasn't important I wouldn't have wasted my time implementing this functionality.

Since iconv is used by SDL we get it for free. Just added additional uchardet for automatic charset detection which isn't something easy to implement, this library is also used by many other text editors like notepad++, etc... which should be easy to meson wrap for static builds on all platforms.

jgmdev commented:
Added a uchardet meson wrap used as fallback in the same static link vein of the other meson wraps. Now the github action CI builds are passing.

Also fixed some lua code mistakes I introduced while synchronizing between the plugin repo changes and this PR.

jgmdev commented:
Now uses official [uchardet wrap](https://github.com/mesonbuild/wrapdb/releases/tag/uchardet_0.0.7-2) thanks to @Jan200101 who submitted a turing complete meson build script to wrapdb.

adamharrison commented:
Tentatively closed, this functionality is available either as a plugin here: https://github.com/jgmdev/lite-xl-encoding, or in Pragtical: https://github.com/pragtical/pragtical.
